### PR TITLE
Add Newton's Method

### PR DESCRIPTION
**Pull request by mikebauer**
## Description of change
Add Newton's Method to the repository.

## Guide to reproduce test results.
```bash
bazel test //resim/math:newton_solver_test
```
## Checklist:
- [x] I have self-reviewed this change.
- [x] I have tested this change.
- [x] This change is covered by tests that are already landed, or in this PR.

### CODE REVIEW

File: resim/math/newton_solver.hh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+#include "resim/utils/nullable_reference.hh"
+#include "resim/utils/status_value.hh"
+
+namespace resim::math {
+
+template <int DIM>
+using DifferentiableFunction = std::function<Eigen::Matrix<double, DIM, 1>(
+    const Eigen::Matrix<double, DIM, 1> &,
+    NullableReference<Eigen::Matrix<double, DIM, DIM>>)>;
```
Comment (knowlsie) [inline resim/math/newton_solver.hh:19]: I think we should comment what's going on here in more detail (i.e. we're publishing out Jacobian, and briefly why it's a NullableReference, for when we don't want to publish Jacobian)

File: resim/math/newton_solver.hh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    Eigen::JacobiSVD<Mat> svd{
+        jacobian,
+        Eigen::ComputeFullU | Eigen::ComputeFullV};
+    result = result - svd.solve(error);
+  }
+  return MAKE_STATUS("Did not converge in Newton iteration!");
+}
+
+template <typename Callable, typename Derived>
+StatusValue<Eigen::Matrix<double, Derived::RowsAtCompileTime, 1>> newton_solve(
```
Comment (knowlsie) [inline resim/math/newton_solver.hh:45]: Could you comment the reason for the wrapper function here? Other than being able to chuck in lambdas without converting them to differentiable functions, it seems the main thing it achieves is the `static_assert` for catching lazy `auto` usage. But would that not be caught by the construction `DifferentiableFunction<DIM>(fun)` anyway, given this has an explicit output type? (Could be wrong on that, but that's why I want a little more clarity.) 

Happy if this is just a nice piece of syntactic sugar, just want the reasoning/usage to be clear to users.

File: resim/math/newton_solver.hh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#include "resim/utils/status_value.hh"
+
+namespace resim::math {
+
+template <int DIM>
+using DifferentiableFunction = std::function<Eigen::Matrix<double, DIM, 1>(
+    const Eigen::Matrix<double, DIM, 1> &,
+    NullableReference<Eigen::Matrix<double, DIM, DIM>>)>;
+
+template <int DIM>
```
Comment (knowlsie) [inline resim/math/newton_solver.hh:21]: I think we should comment the interface here as well. (e.g. this does root-finding on fun, and tolerance is absolute tolerance)

**Review by knowlsie — CHANGES_REQUESTED** [review]
Looks great! I mostly just want more comments on the interface given this is a pretty fundamental thing, but happy otherwise.

File: resim/math/newton_solver.hh (hunk 1 above)
Comment (mikebauer) [reply resim/math/newton_solver.hh:34]: Good call. I've added some comments with an example here! 😀

File: resim/math/newton_solver.hh (hunk 3 above)
Comment (mikebauer) [reply resim/math/newton_solver.hh]: Good call! I've added some.

File: resim/math/newton_solver.hh (hunk 2 above)
Comment (mikebauer) [reply resim/math/newton_solver.hh:105]: Yeah. This is a bit subtle, but it's to allow lambdas to be passed in directly here. I've added some documentation on this.

**Review by mikebauer — COMMENTED** [review]
Thanks for the comments. I think I've resolved them, but let me know if the new comments make sense. I've also made some tweaks to the tolerance to be in the domain rather than the codomain since I think that's more canonical.

File: resim/math/newton_solver.hh [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+// This alias represents a differentiable function from R^N to R^N.
+// Example signiture:
+//
+// using Vec3 = Eigen::Vector3d;
+// using Mat3 = Eigen::Matrix3d;
+//
+// Vec3 my_three_d_function(const Vec3 &x,
+//                          NullableReference<Mat3> jacobian);
+//
+// In this case, the first argument `x` is the intput to the function, and the
```
[the remark this answers is not in the document]
Comment (mikebauer) [reply resim/math/newton_solver.hh:25]: Good catch. Fixing.

**Pull request merged**